File: java/projects/Calculator.java
package projects;
// PROJECT: Java Calculator CLI | javac Calculator.java && java projects.Calculator
// Build a command-line calculator supporting +, -, *, /, ** (power), % (modulo) and parentheses.

public class Calculator {

    // TODO 1 (Mini): Basic REPL — Scanner reads expressions like "3 + 4", compute and print
    //   - Use Scanner(System.in), loop until "exit"
    //   - Split on space: tokens[0]=left, tokens[1]=op, tokens[2]=right
    //   - Handle: +, -, *, /, % — throw ArithmeticException on division by zero
    //
    // TODO 2 (Intermediate): Recursive descent parser (no eval)
    //   - Tokenize input into List<Token> with types: NUMBER, PLUS, MINUS, STAR, SLASH, LPAREN, RPAREN
    //   - Grammar:
    //       expr   → term (('+' | '-') term)*
    //       term   → factor (('*' | '/') factor)*
    //       factor → NUMBER | '(' expr ')' | '-' factor
    //   - Build an AST and evaluate it
    //
    // TODO 3 (Advanced): Variable assignment
    //   - Parse "x = 5 + 3" → store in HashMap<String, Double>
    //   - Reference variables in subsequent expressions: "x * 2"
    //   - Add readline history with jline library or manual up-arrow simulation

    public static void main(String[] args) {
        System.out.println("TODO: implement calculator");
    }
}

File: kotlin/projects/calculator.kt
// PROJECT: Kotlin Calculator CLI | kotlinc calculator.kt -include-runtime -d calculator.jar && java -jar calculator.jar

// TODO 1 (Mini): readLine() REPL loop, split on whitespace, compute "3 + 4"
//   - Support +, -, *, /, % with when expression
//   - Handle ArithmeticException for division by zero
//
// TODO 2 (Intermediate): Tokenizer + recursive descent parser (no eval)
//   - sealed class Token: Num(val n: Double), Plus, Minus, Star, Slash, LParen, RParen
//   - fun tokenize(input: String): List<Token>
//   - fun parseExpr(tokens: ListIterator<Token>): Double
//
// TODO 3 (Advanced): AST with sealed class
//   - sealed class Expr
//   - data class Num(val value: Double) : Expr()
//   - data class BinOp(val op: Char, val left: Expr, val right: Expr) : Expr()
//   - fun eval(expr: Expr): Double = when(expr) { ... }
//
// TODO 4 (Expert): Variable assignment
//   - Parse "x = 5 + 3" → store in mutableMapOf<String, Double>()
//   - Reference variables in subsequent expressions

fun main() {
    println("TODO: implement Kotlin calculator")
}
